### PR TITLE
perf: Adding uprobe test cases with perf command

### PR DESCRIPTION
**Pull request by nasastry**
Adding uprobe test cases to test the following scenarios
test_uprobe
test_uprobe_return
test_uretprobe_variable

Signed-off-by: Nageswara R Sastry <rnsastry@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by nasastry** [comment]
Test results

```
JOB ID     : 9d07e2594d8142834663ade3cb520a209ff4cc03
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-25T19.14-9d07e25/job.log
 (1/3) perf_uprobe.py:perfUprobe.test_uprobe: PASS (2.90 s)
 (2/3) perf_uprobe.py:perfUprobe.test_uprobe_return: PASS (1.47 s)
 (3/3) perf_uprobe.py:perfUprobe.test_uretprobe_variable: PASS (1.48 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 6.46 s
```

**Comment by nasastry** [comment]
Test results with the latest code changes:

```
JOB ID     : 76ee1c1539509b1c108ff75c22e76d4b54f73dad
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-25T19.49-76ee1c1/job.log
 (1/3) perf_uprobe.py:perfUprobe.test_uprobe: PASS (2.93 s)
 (2/3) perf_uprobe.py:perfUprobe.test_uprobe_return: PASS (1.44 s)
 (3/3) perf_uprobe.py:perfUprobe.test_uretprobe_variable: PASS (1.49 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 6.47 s
```

File: perf/perf_uprobe.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+from avocado import Test
+from avocado import main
+from avocado.utils import build, distro, process
+from avocado.utils.software_manager import SoftwareManager
+
+
+class perfUprobe(Test):
+
+    """
+    perf and Uprobe
```
Comment (PraveenPenguin) [inline perf/perf_uprobe.py]: can you please elaborate here what this test does that will helpful to user

File: perf/perf_uprobe.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        self.report = "perf report --input=/tmp/test.out"
+        self.distro_version = detected_distro.version
+
+    def cmd_verify(self, cmd):
+        return process.run(cmd, shell=True)
+
+    def test_uprobe(self):
+        output = self.cmd_verify(self.cmdProbe + ' /usr/bin/perf main')
+        if 'Added new event' not in output.stderr:
+            self.fail("perf probe of perf main failed")
```
Comment (PraveenPenguin) [inline perf/perf_uprobe.py]: can this string as like 
`perf: probe of main subroutine (customize event) failed` and every where it's applicable

File: perf/perf_uprobe.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        output = self.cmd_verify('perf probe -l')
+        if 'probe_perf:main' not in output.stdout:
+            self.fail("perf probe of 'perf main' not found in list")
+        output = self.cmd_verify('cat /sys/kernel/debug/tracing/uprobe_events')
```
Comment (PraveenPenguin) [inline perf/perf_uprobe.py]: please use avocado utility genio for reading file content

File: perf/perf_uprobe.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        if 'probe_perf' not in output.stdout:
+            self.fail("sysfs file didn't reflect uprobe events")
+        output = self.cmd_verify('perf record -o %s -e probe_perf:main -- '
+                                 'perf list' % self.temp_file)
+        if 'samples' not in output.stderr:
+            self.fail("perf.data file not created")
+        output = self.cmd_verify(self.report)
+        print(output)
```
Comment (PraveenPenguin) [inline perf/perf_uprobe.py]: i think it is left over can you please remove this and please other place as well

File: perf/perf_uprobe.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        self.cmdProbe = "perf probe -x"
+        self.recProbe = "perf record -o %s -e probe_uprobe_test:doit" % self.temp_file
+        self.report = "perf report --input=/tmp/test.out"
+        self.distro_version = detected_distro.version
+
+    def cmd_verify(self, cmd):
+        return process.run(cmd, shell=True)
+
+    def test_uprobe(self):
+        output = self.cmd_verify(self.cmdProbe + ' /usr/bin/perf main')
```
Comment (PraveenPenguin) [inline perf/perf_uprobe.py]: please  `%s` to concatenate string 

can be like `'%s  /usr/bin/perf main' % self.cmdProbe` and eve other place as well

File: perf/perf_uprobe.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            print(output)
+        if 'samples' not in output.stderr:
+            self.fail("perf.data file not created")
+        output = self.cmd_verify(self.report)
+        print(output)
+
+    def tearDown(self):
+        output = self.cmd_verify('perf probe -d \\*')
+        if os.path.isfile(self.temp_file):
+            os.remove(self.temp_file)
```
Comment (PraveenPenguin) [inline perf/perf_uprobe.py]: can this be handle it try expect or use avcoado utility process but try expect will be good option as it os command chance of failing this command is there on that case script will error out

**Comment by nasastry** [comment]
@PraveenPenguin Pushed the changed code.
Test results with the new changed code.
```
JOB ID     : 8c1f37479d8d74bc00156beba3f8ea2f68cd8ceb
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-26T11.15-8c1f374/job.log
 (1/3) perf_uprobe.py:perfUprobe.test_uprobe: PASS (2.93 s)
 (2/3) perf_uprobe.py:perfUprobe.test_uprobe_return: PASS (1.44 s)
 (3/3) perf_uprobe.py:perfUprobe.test_uprobe_variable: PASS (1.47 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 6.43 s
```

**Pull request merged**